aptos-move/framework/aptos-stdlib/sources/hash.spec.move

```diff
@@ -1,32 +1,82 @@
 spec aptos_std::aptos_hash {
-    spec sip_hash_from_value {
-        // TODO: temporary mockup.
+    spec module {
+        /// `spec_sip_hash` is an injective function.
```

Comment:
I don't think we should assume this of SipHash. See [here](https://en.wikipedia.org/wiki/SipHash):

 > That is, SHA is designed so that it is difficult for an attacker to find two messages X and Y such that SHA(X) = SHA(Y), even though anyone may compute SHA(X). SipHash instead guarantees that, having seen X_i and SipHash(X_i, k), an attacker who does not know the key k cannot find (any information about) k or SipHash(Y, k) for any message Y ∉ {X_i} which they have not seen before.

Reply:
If we end up assuming injectiveness (i.e., collision resistance), then one could leverage that to prove security of a digital signature scheme which uses SipHash. Yet such as scheme would actually be vulnerable.

Reply:
A fair point! I will not assume the injectiveness for SipHash. By the way, I have a question about our `sip_hash`. According to the wiki page, `SipHash` takes two arguments such as message X and key k. Where is the key in our system? How often is it changed?

Moreover, is it safe to assume that the other hash functions such as `keccak256`, `sha2_512`, `sha3_512` and `ripemd160` are injective?

Reply:
@zekun000 needed a fast hash for Move `Table` implementation, so he used SipHash but forced the key to be zero. (From looking at the Rust native code and my recollections from months ago...)

This is fine for his use, since he doesn't need collision resistance.

However, it is unfortunate that we exported this in Move under the name `sip_hash`. It would've been clearer to name it `sip_hash_with_key_zero` or something like that.

Reply:
And, yes, the rest are safe. 

Although, `ripemd160`'s "injectiveness" can be broken with probability $1/2^{80}$, which is not so good.

Reply:
i.e., one can find $x_1 \ne x_2$ such that $\mathsf{RIPEMD160}(x_1) = \mathsf{RIPEMD160}(x_2)$ with that probability.